### PR TITLE
Web: Load workbook packages from Gist and disk.

### PR DESCRIPTION
**Pull request by bojanrajkovic**

### CODE REVIEW

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        }
+
+        const hasMatchingUti = (!rawMetadata.uti || rawMetadata.uti.localeCompare(defaultUti, undefined, { sensitivity: "base" }));
+        if (checkUti && hasMatchingUti)
+            throw new Error("Invalid manifest");
+        else
+            delete rawMetadata.uti
+
+        let id = rawMetadata.id;
+        if (id && isUUID.v4(id))
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]: I think it's sufficient to simply check that it's just not null/empty. We definitely don't care what version the GUID may be, or even if it's a GUID. The only reason it's a v4 GUID is because that's all that `System.Guid` will produce via `NewGuid`. We don't even use it anywhere<sup>1</sup>. You can eliminate the `is-uuid` dependency.

_<sup>1.</sup> This originated on the first pass of pages where it was going to be used for referencing. If we ever get back to that, it may not even survive._

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    "macnet45": "mac-xamarinmac-full",
+    "macmobile": "mac-xamarinmac-modern",
+    "android": "android-xamarinandroid",
+    "wpf": "wpf",
+    "console": "console",
+    "dotnetcore": "console-netcore"
+});
+
+function getWorkbookTargetIdFromPlatform(platform: string): string {
+    return manifestPlatformToWorkbookTargetMap.get(platform.toLocaleLowerCase())
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]: These are not locale-specific strings (they're effectively enums). Just use `toLowerCase`.

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        saveableManifest.packages = this.manifest.packages.map(pd => {
+            return !pd ? undefined : { id: pd.packageId, version: pd.identityVersion }
+        }).filter((_: any) => _).toArray()
+        saveableManifest.platforms = this.manifest.platforms.map(wt => {
+            return !wt ? undefined : getPlatformFromWorkbookTargetId(wt.id)
+        }).filter((_: any) => _).toArray()
+        return saveableManifest
+    }
+
+    parseManifest(rawMetadata: any, workbookSession: WorkbookSession): WorkbookManifest {
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]: Would be good to explain the general idea of how this whole method works. It's a little hard to follow with all of the deleting properties out of `rawMetadata`. Should also explain that it will round-trip any and all arbitrary YAML and just pulls out bits that we actually need.

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        const platform: string = rawMetadata.platform;
+        const platforms: string[] = rawMetadata.platforms || [];
+        finalMetadata.platforms = List(platforms.concat(platform).filter(_ => _).map(workbookPlatform => {
+            const matchingTarget = workbookSession.availableWorkbookTargets.find(value => {
+                return value.id === getWorkbookTargetIdFromPlatform(workbookPlatform)
+            })
+            if (!matchingTarget)
+                console.warn(`Could not find matching target for workbook platform ${workbookPlatform}.`)
+            return matchingTarget;
+        }).filter(_ => _))
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]: This call to `.filter` seems entirely pointless. Explain?

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/components/WorkbookShell.tsx [hunk 5]
```diff
@@ -138,26 +140,61 @@ export class WorkbookShell extends React.Component<any, WorkbookShellState> {
         this.fileButton.click();
     }
 
-    loadWorkbook(event: React.ChangeEvent<HTMLInputElement>) {
+    async triggerGistPicker() {
+        if (!this.workbookEditor)
+            return;
+
+        const gistUrl = prompt("Enter the URL of the Gist you want to load from?");
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/components/WorkbookShell.tsx]: `// TODO: real UI` Hello 1996!

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/components/WorkbookShell.tsx [hunk 6]
```diff
@@ -138,26 +140,61 @@ export class WorkbookShell extends React.Component<any, WorkbookShellState> {
...
         const file = event.target.files[0]
         const reader = new FileReader
-        reader.addEventListener('load', () => this.restorePackages (reader))
-        reader.readAsText(file)
+        reader.addEventListener('load', () => this.loadWorkbook(file, reader))
+        reader.readAsArrayBuffer(file)
+    }
+
+    sniffHeader(file: ArrayBuffer): Boolean {
+        const magic = new Uint32Array(file.slice(0, 4))[0]
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/components/WorkbookShell.tsx]: Bounds check? What if 3 byte file?

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/typings/custom.d.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
+declare module "draft-js-plugins-editor";
+declare module "draft-js-markdown-plugin";
+declare module "turndown";
+declare module "is-uuid";
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/typings/custom.d.ts]: Nuke.

File: Clients/Xamarin.Interactive.Client.Web/Controllers/GistController.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
+//
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/Controllers/GistController.cs]: This entire class needs formatting per our .editorconfig.

File: Clients/Xamarin.Interactive.Client.Web/package.json [hunk 9]
```diff
@@ -47,6 +48,8 @@
     "file-saver": "^1.3.3",
     "gray-matter": "^3.1.1",
     "immutable": "^3.8.2",
+    "is-uuid": "^1.0.2",
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/package.json]: Nuke

File: Clients/Xamarin.Interactive.Client.Web/Controllers/GistController.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+                var gist = JObject.Parse(await gistApiResponse.Content.ReadAsStringAsync());
+                var latestGistRevision = (string)gist["history"][0]["version"];
+                var gistOwner = (string)gist["owner"]["login"];
+
+                var gistZipUrl = $"https://gist.github.com/{gistOwner}/{id}/archive/{latestGistRevision}.zip";
+                var gistZipResponse = await hc.GetAsync(gistZipUrl);
+                if (!gistZipResponse.IsSuccessStatusCode)
+                    return StatusCode((int)gistZipResponse.StatusCode);
+
+                return File(await gistZipResponse.Content.ReadAsStreamAsync(), "application/zip", $"{latestGistRevision}.zip");
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/Controllers/GistController.cs]: Wrap 3 lines:

```csharp
return File (
    await gistZipResponse.Content.ReadAsStreamAsync (),
    "application/zip",
    $"{latestGistRevision}.zip");
```

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts (hunk 2 above)
Comment (bojanrajkovic) [reply Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]: 👍 Looks like bad autocomplete from VS Code. Never even _seen_ that method used.

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts (hunk 4 above)
Comment (bojanrajkovic) [reply Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]: It's filtering out undefined values from the final list--if the rest of the code can deal with undefined values in the workbook targets, that can be removed.

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/components/WorkbookShell.tsx (hunk 6 above)
Comment (bojanrajkovic) [reply Clients/Xamarin.Interactive.Client.Web/ClientApp/components/WorkbookShell.tsx]: 👍 I had tested this and Uint32Array didn't seem to care, but I tested wrong, it does.

File: Clients/Xamarin.Interactive.Client.Web/Controllers/GistController.cs (hunk 8 above)
Comment (bojanrajkovic) [reply Clients/Xamarin.Interactive.Client.Web/Controllers/GistController.cs]: Weird, I have the EditorConfig plugin for VSCode...

File: Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        // doesn't know about, without duplicating any of it. We'll remove properties we actually consume
+        // from the `rawMetadata` object (all of the deletes below) so that we can store it as the
+        // preserved metadata.
+        if (!rawMetadata) {
+            rawMetadata = {}
+            checkUti = false
+        } else
+            rawMetadata = Object.assign({}, rawMetadata)
+
+        const hasMatchingUti = (!rawMetadata.uti || rawMetadata.uti.localeCompare(defaultUti, undefined, { sensitivity: "base" }));
```
Comment (abock) [inline Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts]:
```javascript
const hasMatchingUti = !rawMetadata.uti || rawMetadata.uti.toLower() === defaultUti
```
